pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+	}
+	// Update mesh for new label mappings and new subset mappings
+	m.reverseAttrMap.addAttributeValues(newLabelMappings)
+	m.subsetEndpoints.addSubsetEndpoints(newSubsetMappings)
+	return nil
+}
+
+// MeshEndpoints implements functionality required for EDS and returns a list of endpoints
+// that match one or more subsets.
+func (m *Mesh) MeshEndpoints(subsetNames []string) []*Endpoint {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	epSet := endpointUIDSet{}
+	for _, ssName := range subsetNames {
+		ssEpSet, found := m.subsetEndpoints[ssName]
+		if !found {
+			continue
+		}
+		epSet.appendAll(ssEpSet)
+	}
+	out := make([]*Endpoint, len(epSet))
+	idx := 0
+	for uid, _ := range epSet {
+		out[idx] = m.allEndpoints[uid]
+		idx++
+	}
+	return out
+}
+
+// SubsetNames implements functionality required for CDS and returns a list of all subset
+// names currently configured for this Mesh
+func (m *Mesh) SubsetNames() []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	out := make([]string, len(m.subsetEndpoints))
+	idx := 0
+	for ssName := range m.subsetEndpoints {
+		out[idx] = ssName
+		idx++
+	}
+	return out
+}
+
+// UpdateSubsets implements functionality required for pilot configuration (via Galley).
+// It updates the Mesh for supplied events, adding, updating and deleting Subsets
+// from this mesh as determined by the corresponding Event. The list of subsetEvents
+// can be a partial one.
+func (m *Mesh) UpdateSubsets(subsetEvents []SubsetEvent) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, ssEvent := range subsetEvents {
+		subset := ssEvent.Subset
+		if ssEvent.Event == EventDelete || ssEvent.Event == EventUpdate {
+			ssName := subset.Name
+			oldEpKeySet := m.subsetEndpoints[ssName]
+			for epKey := range oldEpKeySet {
+				m.reverseEpSubsets.deleteSubset(epKey, ssName)
+			}
+			delete(m.subsetEndpoints, ssName)
+		}
+		if ssEvent.Event == EventAdd || ssEvent.Event == EventUpdate {
+			ssName := subset.Name
+			epUIDSet := m.reverseAttrMap.getKeysMatching(subset.Labels)
+			m.subsetEndpoints.addEndpointUIDs(ssName, epUIDSet)
+			for endpointUID := range epUIDSet {
+				m.reverseEpSubsets.addSubset(endpointUID, ssName)
+			}
+		}
+	}
+}
+
+// NewEndpoint is a boiler plate function intended for platform Controllers to create a new Endpoint.
+// This method ensures all the necessary data required for creating subsets are correctly setup. It
+// also performs sorting of arrays etc, to allow stable results for reflect.DeepEquals() for quick
+// comparisons.
+func NewEndpoint(endpointUID, service string, aliases []string, address string, port uint32, protocol Protocol, labels Labels, user string) (*Endpoint, error) {
+	var errs error
+	ipAddr := net.ParseIP(address)
+	if ipAddr == nil {
+		errs = multierror.Append(errs, errors.New(fmt.Sprintf("invalid IP address '%s'", address)))
+	}
+	svcParts := svcRegexp.FindStringSubmatch(service)
+	if svcParts == nil || len(svcParts) != 3 {
+		errs = multierror.Append(
+			errors.New(fmt.Sprintf(
```

Comment:
Golang style has no line limit lengths. When you change this to `fmt.Errorf`, keep the entire error string on one line to make it more readable.